staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -56,7 +57,20 @@ func newMutatingDispatcher(p *Plugin) func(cm *webhook.ClientManager) generic.Di
 var _ generic.Dispatcher = &mutatingDispatcher{}
 
 func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr *generic.VersionedAttributes, o admission.ObjectInterfaces, relevantHooks []*v1beta1.Webhook) error {
+	reinvokeCtx := attr.GetReinvocationContext()
+	if reinvokeCtx.IsReinvoke() && reinvokeCtx.IsOutputChangedSinceLastWebhookInvocation(attr.Attributes.GetObject()) {
+		// If the object has changed, we know the in-tree plugin re-invocations have mutated the object,
+		// and we need to reinvoke all eligible webhooks.
+		reinvokeCtx.RequireReinvokingPreviouslyInvokedPlugins()
+	}
+	defer func() {
+		reinvokeCtx.SetLastWebhookInvocationOutput(attr.Attributes.GetObject())
+	}()
+
 	for _, hook := range relevantHooks {
+		if !reinvokeCtx.ShouldInvokeWebhook(hook.Name) {
```

Comment:
why this complicated logic keeping the string set up to date in the context and then checking it here on reinvoke? As far as I see one only has to check `ReinvocationPolicy==IfNeeded` on reinvoke here, no need for state. What do I miss?

Reply:
the string sets buffer `Policy==IfNeeded` hooks, and only flush the buffer to "real invocation" set if a later plugin mutated the object. This avoids unnecessary invocation given hooks are idempotent

Reply:
See discussion in https://github.com/kubernetes/kubernetes/pull/78080#discussion_r286128070

Reply:
the state also helps us in the case where new webhook gets configured during the second pass. If we only depend on the `ReinvocationPolicy` bit, a new webhook could race the second pass and get invoked once. With the state we make sure that a webhook invoked in the second pass must have been invoked in the first pass before